File: server/src/frameworks/load.ts
/**
 * Typed, cached views over data/frameworks/*.json (ASVS 5.0.0, AISVS 1.0, AISVS Appendix C, SbD 0.5.0).
 *
 * Ids: ASVS "V6.2.1" (chapter V6, section V6.2); AISVS "C2.1.3" (chapter C2, section C2.1);
 * Appendix C "AC.4.1" (family AC.4, which doubles as chapter and section); SbD "AS-01".
 */
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { z } from 'zod';
import type { StandardId } from '@shared/compliance.js';

export const REPO_ROOT = fileURLToPath(new URL('../../..', import.meta.url));
export const DATA_DIR = join(REPO_ROOT, 'data');
export const FRAMEWORKS_DIR = join(DATA_DIR, 'frameworks');
export const KNOWLEDGE_DIR = join(DATA_DIR, 'knowledge');

const RequirementJsonSchema = z.object({
  id: z.string(),
  description: z.string(),
  level: z.number().int(),
  cwe: z.array(z.union([z.string(), z.number()])).optional(),
  mappings: z.string().optional(),
});

const AsvsJsonSchema = z.object({
  standard: z.string(),
  version: z.string(),
  source: z.string().optional(),
  license: z.string().optional(),
  chapters: z.array(
    z.object({
      id: z.string(),
      shortName: z.string().optional(),
      name: z.string(),
      sections: z.array(z.object({ id: z.string(), name: z.string(), requirements: z.array(RequirementJsonSchema) })),
    }),
  ),
});

const AisvsJsonSchema = z.object({
  standard: z.string(),
  version: z.string(),
  source: z.string().optional(),
  license: z.string().optional(),
  chapters: z.array(
    z.object({
      id: z.string(),
      name: z.string(),
      controlObjective: z.string().optional(),
      sections: z.array(
        z.object({
          id: z.string(),
          name: z.string(),
          description: z.string().optional(),
          requirements: z.array(RequirementJsonSchema),
        }),
      ),
    }),
  ),
});

const AppendixCJsonSchema = z.object({
  standard: z.string(),
  version: z.string(),
  source: z.string().optional(),
  license: z.string().optional(),
  families: z.array(
    z.object({
      id: z.string(),
      name: z.string(),
      description: z.string().optional(),
      requirements: z.array(RequirementJsonSchema),
    }),
  ),
});

const SbdJsonSchema = z.object({
  framework: z.string(),
  version: z.string(),
  source: z.string().optional(),
  license: z.string().optional(),
  checklistDomains: z.array(
    z.object({
      id: z.string(),
      name: z.string(),
      controls: z.array(
        z.object({
          id: z.string(),
          statement: z.string(),
          critical: z.boolean(),
          severityIfNo: z.enum(['high', 'medium', 'low']),
        }),
      ),
    }),
  ),
  scoring: z.object({
    yes: z.number(),
    'n-a': z.number(),
    no: z.object({ low: z.number(), medium: z.number(), high: z.number() }),
    escalateWhen: z.string(),
    threshold: z.number(),
  }),
  processSteps: z.array(z.string()),
  principles: z.array(z.object({ id: z.string(), name: z.string() })),
  escalationTriggers: z.array(z.object({ id: z.string(), name: z.string() })),
});

export interface RequirementInfo {
  id: string;
  standard: StandardId;
  chapterId: string;
  chapterName: string;
  sectionId: string;
  sectionName: string;
  description: string;
  level: number;
}

export interface SectionInfo {
  id: string;
  name: string;
  requirementIds: string[];
}

export interface ChapterInfo {
  id: string;
  name: string;
  shortName?: string;
  standard: StandardId;
  sections: SectionInfo[];
}

export interface StandardView {
  standard: StandardId;
  name: string;
  version: string;
  source?: string;
  license?: string;
  chapters: ChapterInfo[];
  requirements: RequirementInfo[];
}

export interface SbdControlInfo {
  id: string;
  domain: string; // "A".."E"
  domainName: string;
  statement: string;
  critical: boolean;
  severityIfNo: 'high' | 'medium' | 'low';
}

export interface SbdView {
  framework: string;
  version: string;
  source?: string;
  license?: string;
  domains: { id: string; name: string; controlIds: string[] }[];
  controls: SbdControlInfo[];
  scoring: z.infer<typeof SbdJsonSchema>['scoring'];
  processSteps: string[];
  principles: { id: string; name: string }[];
  escalationTriggers: { id: string; name: string }[];
}

export interface Frameworks {
  asvs: StandardView;
  aisvs: StandardView;
  appendixC: StandardView;
  sbd: SbdView;
  /** Fast lookup for requirement ids of any standard (V6.2.1, C2.1.3, AC.4.1). */
  getRequirement(id: string): RequirementInfo | undefined;
  listRequirements(standard: StandardId): RequirementInfo[];
  chapters(standard: StandardId): ChapterInfo[];
  getSbdControl(id: string): SbdControlInfo | undefined;
  /** True when the id names a known chapter, section or requirement of any standard. */
  hasScope(id: string): boolean;
  /** Requirement ids under a chapter/section/requirement id (empty when unknown). */
  requirementIdsInScope(scope: string): string[];
}

function readJson(file: string): unknown {
  return JSON.parse(readFileSync(file, 'utf8')) as unknown;
}

function buildStandard(
  standard: StandardId,
  meta: { standard: string; version: string; source?: string; license?: string },
  chapters: {
    id: string;
    name: string;
    shortName?: string;
    sections: { id: string; name: string; requirements: z.infer<typeof RequirementJsonSchema>[] }[];
  }[],
): StandardView {
  const requirements: RequirementInfo[] = [];
  const chapterInfos: ChapterInfo[] = chapters.map((ch) => ({
    id: ch.id,
    name: ch.name,
    ...(ch.shortName ? { shortName: ch.shortName } : {}),
    standard,
    sections: ch.sections.map((s) => {
      for (const r of s.requirements) {
        requirements.push({
          id: r.id,
          standard,
          chapterId: ch.id,
          chapterName: ch.name,
          sectionId: s.id,
          sectionName: s.name,
          description: r.description,
          level: r.level,
        });
      }
      return { id: s.id, name: s.name, requirementIds: s.requirements.map((r) => r.id) };
    }),
  }));
  return {
    standard,
    name: meta.standard,
    version: meta.version,
    ...(meta.source ? { source: meta.source } : {}),
    ...(meta.license ? { license: meta.license } : {}),
    chapters: chapterInfos,
    requirements,
  };
}

let cache: Frameworks | undefined;

/** Loads and caches the four framework files. Pass a directory only in tests. */
export function loadFrameworks(frameworksDir: string = FRAMEWORKS_DIR): Frameworks {
  if (cache && frameworksDir === FRAMEWORKS_DIR) return cache;

  const asvsJson = AsvsJsonSchema.parse(readJson(join(frameworksDir, 'asvs-5.0.0.json')));
  const aisvsJson = AisvsJsonSchema.parse(readJson(join(frameworksDir, 'aisvs-1.0.json')));
  const appendixJson = AppendixCJsonSchema.parse(readJson(join(frameworksDir, 'aisvs-1.0-appendix-c.json')));
  const sbdJson = SbdJsonSchema.parse(readJson(join(frameworksDir, 'sbd-checklist-0.5.0.json')));

  const asvs = buildStandard('asvs', asvsJson, asvsJson.chapters);
  const aisvs = buildStandard('aisvs', aisvsJson, aisvsJson.chapters);
  // Appendix C has families only: each family is both the chapter and the section.
  const appendixC = buildStandard(
    'aisvs-appendix-c',
    appendixJson,
    appendixJson.families.map((f) => ({
      id: f.id,
      name: f.name,
      sections: [{ id: f.id, name: f.name, requirements: f.requirements }],
    })),
  );

  const controls: SbdControlInfo[] = [];
  const domains = sbdJson.checklistDomains.map((d) => {
    for (const c of d.controls) controls.push({ ...c, domain: d.id, domainName: d.name });
    return { id: d.id, name: d.name, controlIds: d.controls.map((c) => c.id) };
  });
  const sbd: SbdView = {
    framework: sbdJson.framework,
    version: sbdJson.version,
    ...(sbdJson.source ? { source: sbdJson.source } : {}),
    ...(sbdJson.license ? { license: sbdJson.license } : {}),
    domains,
    controls,
    scoring: sbdJson.scoring,
    processSteps: sbdJson.processSteps,
    principles: sbdJson.principles,
    escalationTriggers: sbdJson.escalationTriggers,
  };

  const byStandard: Record<StandardId, StandardView> = { asvs, aisvs, 'aisvs-appendix-c': appendixC };
  const requirementIndex = new Map<string, RequirementInfo>();
  const scopeIndex = new Map<string, string[]>();
  for (const view of Object.values(byStandard)) {
    for (const r of view.requirements) requirementIndex.set(r.id, r);
    for (const ch of view.chapters) {
      scopeIndex.set(
        ch.id,
        ch.sections.flatMap((s) => s.requirementIds),
      );
      for (const s of ch.sections) scopeIndex.set(s.id, [...s.requirementIds]);
    }
  }
  const controlIndex = new Map(controls.map((c) => [c.id, c]));

  const frameworks: Frameworks = {
    asvs,
    aisvs,
    appendixC,
    sbd,
    getRequirement: (id) => requirementIndex.get(id),
    listRequirements: (standard) => byStandard[standard].requirements,
    chapters: (standard) => byStandard[standard].chapters,
    getSbdControl: (id) => controlIndex.get(id),
    hasScope: (id) => requirementIndex.has(id) || scopeIndex.has(id),
    requirementIdsInScope: (scope) => {
      if (requirementIndex.has(scope)) return [scope];
      return scopeIndex.get(scope) ?? [];
    },
  };
  if (frameworksDir === FRAMEWORKS_DIR) cache = frameworks;
  return frameworks;
}

/** Which standard an id belongs to, by its shape (does not check existence). */
export function standardForId(id: string): StandardId | 'sbd' | undefined {
  if (/^AC\.\d+(\.\d+)?$/.test(id)) return 'aisvs-appendix-c';
  if (/^V\d+(\.\d+){0,2}$/.test(id)) return 'asvs';
  if (/^C\d+(\.\d+){0,2}$/.test(id)) return 'aisvs';
  if (/^(AS|DM|RR|AC|MT)-\d\d$/.test(id)) return 'sbd';
  return undefined;
}

/** The section id of a requirement id ("V6.2.1" → "V6.2", "AC.4.1" → "AC.4"). */
export function sectionIdOf(requirementId: string): string {
  return requirementId.slice(0, requirementId.lastIndexOf('.'));
}

/** The chapter id of a requirement or section id ("V6.2.1" → "V6", "AC.4.1" → "AC.4"). */
export function chapterIdOf(id: string): string {
  if (id.startsWith('AC.')) {
    const parts = id.split('.');
    return `${parts[0]}.${parts[1]}`;
  }
  return id.split('.')[0] ?? id;
}

File: agnostic/crates/sv-frameworks/src/load.rs
//! Loads the OWASP data files. These are the same files v1 reads, from the same `data/frameworks`
//! folder: one source of truth for ASVS across both products, so a correction fixes both.
//!
//! ASVS and AISVS share a shape (chapters → sections → requirements). Appendix C is families →
//! requirements, with no section level; its ids carry two segments of family ("AC.4.1" → "AC.4"),
//! which is why `chapter_id_of` special-cases them.

use anyhow::{Context, Result};
use serde::Deserialize;
use std::collections::BTreeMap;
use std::path::Path;

#[derive(Debug, Deserialize)]
pub struct Requirement {
    pub id: String,
    pub description: String,
    pub level: u8,
}

#[derive(Debug, Deserialize)]
pub struct Section {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub requirements: Vec<Requirement>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Chapter {
    pub id: String,
    pub name: String,
    /// ASVS and AISVS nest a section level; Appendix C families do not.
    #[serde(default)]
    pub sections: Vec<Section>,
    /// Appendix C families hold requirements directly.
    #[serde(default)]
    pub requirements: Vec<Requirement>,
}

/// Namespace for the Secure by Design checklist's control ids. See `load_checklist`.
pub const SBD_PREFIX: &str = "SBD-";

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ChecklistFile {
    checklist_domains: Vec<ChecklistDomain>,
}

#[derive(Debug, Deserialize)]
struct ChecklistDomain {
    id: String,
    name: String,
    controls: Vec<ChecklistControl>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ChecklistControl {
    id: String,
    statement: String,
    #[serde(default)]
    critical: bool,
    #[serde(default)]
    severity_if_no: String,
}

/// The checklist has no levels, so one is derived from how much its absence costs.
///
/// This is SecureVibe's mapping and not OWASP's, which is why it is one function with a name rather
/// than three comparisons spread around.
fn level_of(control: &ChecklistControl) -> u8 {
    if control.critical || control.severity_if_no == "high" {
        1
    } else if control.severity_if_no == "medium" {
        2
    } else {
        3
    }
}

#[derive(Debug, Deserialize)]
pub struct FrameworkFile {
    #[serde(default)]
    pub standard: String,
    #[serde(default)]
    pub version: String,
    /// ASVS and AISVS call the top level `chapters`; Appendix C calls it `families`.
    #[serde(default, alias = "families")]
    pub chapters: Vec<Chapter>,
}

/// One requirement with the context needed to reason about it, flattened out of the nesting.
#[derive(Debug, Clone)]
pub struct RequirementInfo {
    pub id: String,
    pub description: String,
    pub level: u8,
    pub chapter_id: String,
    pub chapter_name: String,
    /// Absent for Appendix C, which has no section level.
    pub section_id: Option<String>,
    /// Where the level came from, when it is not the framework's own: the Secure by Design checklist
    /// has no levels, so each of its controls says how it got one. `None` for ASVS and AISVS.
    pub level_basis: Option<String>,
    /// Requirements in another framework that ask the same thing, from the crosswalk. Evidence about
    /// one of these is shown beside this requirement as supporting, never as checking it.
    pub counterparts: Vec<String>,
}

#[derive(Debug, Default)]
pub struct Frameworks {
    /// Every requirement from every loaded file, keyed by id.
    pub requirements: BTreeMap<String, RequirementInfo>,
}

impl Frameworks {
    /// Loads ASVS, AISVS and Appendix C from a `data/frameworks` directory.
    pub fn load(frameworks_dir: &Path) -> Result<Self> {
        let mut out = Self::default();
        for file in [
            "asvs-5.0.0.json",
            "aisvs-1.0.json",
            "aisvs-1.0-appendix-c.json",
        ] {
            out.load_file(&frameworks_dir.join(file))?;
        }
        out.load_checklist(&frameworks_dir.join("sbd-checklist-0.5.0.json"))?;
        Ok(out)
    }

    fn load_file(&mut self, path: &Path) -> Result<()> {
        let text = std::fs::read_to_string(path)
            .with_context(|| format!("reading framework file {}", path.display()))?;
        let parsed: FrameworkFile = serde_json::from_str(&text)
            .with_context(|| format!("parsing framework file {}", path.display()))?;
        for chapter in &parsed.chapters {
            for req in &chapter.requirements {
                self.insert(req, chapter, None);
            }
            for section in &chapter.sections {
                for req in &section.requirements {
                    self.insert(req, chapter, Some(&section.id));
                }
            }
        }
        Ok(())
    }

    /// The Secure by Design checklist, which is a third schema and not a set of requirements.
    ///
    /// `sv --help` and the README claimed this was checked for as long as it was not loaded at all,
    /// which is the plainest kind of untruth this tool can tell. Two things had to be decided before
    /// it could be loaded, and both are choices rather than readings:
    ///
    /// **Ids are prefixed `SBD-`.** The checklist names its own controls `AC-01` … `AC-07`, and
    /// AISVS Appendix C already owns `AC.1.1` … `AC.13.4`. Those are different strings and would not
    /// collide in the map, but they collide in a reader's head — and this codebase has already
    /// shipped five checkers citing `AC-05` for a family written `AC.5`. A citation that resolves to
    /// the wrong framework is worse than one that resolves to nothing, because nothing looks wrong.
    /// So the checklist's ids are namespaced here and the prefix is printed everywhere they appear.
    ///
    /// **Levels are derived, because the checklist has none.** It has `critical` and `severityIfNo`
    /// instead. A control that is critical, or whose absence is high severity, is level 1; medium is
    /// level 2; low is level 3. That mapping is this tool's, not OWASP's, and is stated wherever the
    /// level is shown.
    fn load_checklist(&mut self, path: &Path) -> Result<()> {
        let text = std::fs::read_to_string(path)
            .with_context(|| format!("reading checklist file {}", path.display()))?;
        let parsed: ChecklistFile = serde_json::from_str(&text)
            .with_context(|| format!("parsing checklist file {}", path.display()))?;
        for domain in &parsed.checklist_domains {
            let chapter = Chapter {
                id: format!("{SBD_PREFIX}{}", domain.id),
                name: domain.name.clone(),
                sections: Vec::new(),
                requirements: Vec::new(),
            };
            for control in &domain.controls {
                let requirement = Requirement {
                    id: format!("{SBD_PREFIX}{}", control.id),
                    description: control.statement.clone(),
                    level: level_of(control),
                };
                self.insert(&requirement, &chapter, None);
                if let Some(info) = self.requirements.get_mut(&requirement.id) {
                    info.level_basis = Some(format!(
                        "level {}, set by sv from the checklist's severity; the checklist has no levels",
                        requirement.level
                    ));
                }
            }
        }
        Ok(())
    }

    /// Grounds the checklist's levels in ASVS through the crosswalk in `path`.
    ///
    /// A control's level becomes the lower of the one derived from its severity and the lowest level
    /// among the ASVS requirements that ask the same thing. Lower only: the crosswalk can bring a
    /// control into an app's scope sooner and can never take one out. A control nothing in ASVS
    /// matches is level 1 — shown at every target — because excluding it would rest on `sv`'s own
    /// derived level and nothing else. Every control has to be listed, and every id cited has to
    /// exist; either failing is a load error rather than a control quietly keeping its old level.
    pub fn apply_crosswalk(&mut self, path: &Path) -> Result<()> {
        #[derive(Deserialize)]
        struct Crosswalk {
            /// Control → ASVS requirement → the few words naming what the two ask in common.
            controls: BTreeMap<String, BTreeMap<String, String>>,
        }
        let text = std::fs::read_to_string(path)
            .with_context(|| format!("reading crosswalk {}", path.display()))?;
        let crosswalk: Crosswalk = serde_json::from_str(&text)
            .with_context(|| format!("parsing crosswalk {}", path.display()))?;

        let controls: Vec<String> = self
            .requirements
            .keys()
            .filter(|id| id.starts_with(SBD_PREFIX))
            .cloned()
            .collect();
        for id in &controls {
            anyhow::ensure!(
                crosswalk.controls.contains_key(id),
                "the crosswalk does not list {id}; every control has to be listed, with an empty list \
                 when nothing in ASVS asks the same thing"
            );
        }
        for (id, counterparts) in &crosswalk.controls {
            anyhow::ensure!(
                self.requirements.contains_key(id),
                "the crosswalk lists {id}, which is not a loaded control"
            );
            let mut lowest: Option<(u8, &str)> = None;
            for asvs in counterparts.keys() {
                let level = self
                    .requirements
                    .get(asvs)
                    .with_context(|| {
                        format!("the crosswalk cites {asvs} for {id}, which does not exist")
                    })?
                    .level;
                if lowest.is_none_or(|(l, _)| level < l) {
                    lowest = Some((level, asvs));
                }
            }
            let info = self.requirements.get_mut(id).expect("checked above");
            let derived = info.level;
            info.counterparts = counterparts.keys().cloned().collect();
            let (level, basis) = match lowest {
                None => (
                    1,
                    "shown at every level: nothing in ASVS asks the same thing, and the checklist has \
                     no levels"
                        .to_owned(),
                ),
                Some((asvs_level, asvs)) if asvs_level <= derived => {
                    (asvs_level, format!("level {asvs_level}, as {asvs}"))
                }
                Some((asvs_level, asvs)) => (
                    derived,
                    format!(
                        "level {derived}, from the checklist's severity; its ASVS counterpart {asvs} \
                         is level {asvs_level}"
                    ),
                ),
            };
            info.level = level;
            info.level_basis = Some(basis);
        }
        Ok(())
    }

    fn insert(&mut self, req: &Requirement, chapter: &Chapter, section_id: Option<&str>) {
        self.requirements.insert(
            req.id.clone(),
            RequirementInfo {
                id: req.id.clone(),
                description: req.description.clone(),
                level: req.level,
                chapter_id: chapter.id.clone(),
                chapter_name: chapter.name.clone(),
                section_id: section_id.map(str::to_owned),
                level_basis: None,
                counterparts: Vec::new(),
            },
        );
    }

    pub fn get(&self, id: &str) -> Option<&RequirementInfo> {
        self.requirements.get(id)
    }

    pub fn len(&self) -> usize {
        self.requirements.len()
    }

    pub fn is_empty(&self) -> bool {
        self.requirements.is_empty()
    }
}

/// The section id of a requirement id ("V6.2.1" → "V6.2").
pub fn section_id_of(requirement_id: &str) -> &str {
    // The checklist has no section level, so a control's section is its family. Without this the
    // whole id comes back and a rule scoped to a family would never match one.
    if requirement_id.starts_with(SBD_PREFIX) {
        return chapter_id_of(requirement_id);
    }
    match requirement_id.rfind('.') {
        Some(i) => &requirement_id[..i],
        None => requirement_id,
    }
}

/// The chapter id of a requirement or section id ("V6.2.1" → "V6", "AC.4.1" → "AC.4").
pub fn chapter_id_of(id: &str) -> &str {
    // "SBD-AC-01" -> "SBD-AC": the family, so one rule can scope every access-control control.
    if let Some(after_prefix) = id.strip_prefix(SBD_PREFIX) {
        let end = after_prefix
            .find('-')
            .map_or(id.len(), |i| SBD_PREFIX.len() + i);
        return &id[..end];
    }
    if let Some(rest) = id.strip_prefix("AC.") {
        // Appendix C ids carry two segments of family, so keep "AC." plus the next one.
        let end = rest.find('.').map_or(id.len(), |i| "AC.".len() + i);
        return &id[..end];
    }
    match id.find('.') {
        Some(i) => &id[..i],
        None => id,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn derives_section_and_chapter_ids() {
        assert_eq!(section_id_of("V6.2.1"), "V6.2");
        assert_eq!(chapter_id_of("V6.2.1"), "V6");
        assert_eq!(chapter_id_of("V6.2"), "V6");
        assert_eq!(chapter_id_of("AC.4.1"), "AC.4");
        assert_eq!(chapter_id_of("AC.4"), "AC.4");
        assert_eq!(chapter_id_of("C9.2.1"), "C9");
        // The checklist. "SBD-AC-01" must not be read as Appendix C's "AC." family, and its
        // section has to be its family or no family-scoped rule would ever match.
        assert_eq!(chapter_id_of("SBD-AC-01"), "SBD-AC");
        assert_eq!(chapter_id_of("SBD-MT-07"), "SBD-MT");
        assert_eq!(section_id_of("SBD-AC-01"), "SBD-AC");
    }

    #[test]
    fn the_checklist_level_is_derived_from_what_its_absence_costs() {
        let critical = ChecklistControl {
            id: "AS-01".into(),
            statement: String::new(),
            critical: true,
            severity_if_no: "high".into(),
        };
        assert_eq!(level_of(&critical), 1);
        let high_but_not_critical = ChecklistControl {
            severity_if_no: "high".into(),
            critical: false,
            ..ChecklistControl {
                id: "AC-03".into(),
                statement: String::new(),
                critical: false,
                severity_if_no: "high".into(),
            }
        };
        assert_eq!(level_of(&high_but_not_critical), 1, "high alone is enough");
        let medium = ChecklistControl {
            id: "DM-01".into(),
            statement: String::new(),
            critical: false,
            severity_if_no: "medium".into(),
        };
        assert_eq!(level_of(&medium), 2);
        let low = ChecklistControl {
            id: "AS-02".into(),
            statement: String::new(),
            critical: false,
            severity_if_no: "low".into(),
        };
        assert_eq!(level_of(&low), 3);
    }
}
